File: test/differential/oracle.py
"""Persistent differential-test oracle: run generated xarray code against *real* xarray.

Opens the dumped store **once** as ``ds``, then reads line-delimited JSON requests
from stdin — each ``{"id", "code"}`` where ``code`` is the transform tail xarray-ts
emitted (e.g. ``ds.isel(time=0)["temperature"]``). It ``eval``s that against the
open dataset and replies, one JSON line per request, with the resulting structure
and a per-data-variable value signature. Evaluating the emitted string is the
point: it exercises the actual codegen output, not a re-interpretation of it.

Coordinates are deliberately excluded from the value signature — xarray-ts decodes
CF time to epoch-ms while xarray uses ``datetime64``, so coord *values* wouldn't
compare across the boundary (dims/names still do).

Usage: ``python oracle.py <store-dir>``.
"""

import json
import sys

import numpy as np
import xarray as xr


def sig_of(da: xr.DataArray) -> dict:
    """A cheap, comparable value signature for one array: shape + sum/first/last."""
    arr = np.asarray(da.values, dtype="float64").ravel()
    if arr.size == 0:
        return {"shape": list(da.shape), "sum": None, "first": None, "last": None}
    return {
        "shape": list(da.shape),
        "sum": float(np.nan_to_num(arr).sum()),
        "first": float(arr[0]),
        "last": float(arr[-1]),
    }


def describe(result) -> dict:
    """Structure + value signature of an eval result (a Dataset or a DataArray)."""
    if isinstance(result, xr.Dataset):
        return {
            "kind": "dataset",
            "dims": {str(k): int(v) for k, v in result.sizes.items()},
            "data_vars": sorted(map(str, result.data_vars)),
            "coords": sorted(map(str, result.coords)),
            "sig": {str(k): sig_of(result[k]) for k in result.data_vars},
        }
    if isinstance(result, xr.DataArray):
        return {
            "kind": "dataarray",
            "name": None if result.name is None else str(result.name),
            "dims": {str(k): int(v) for k, v in result.sizes.items()},
            "coords": sorted(map(str, result.coords)),
            "sig": sig_of(result),
        }
    raise TypeError(f"unexpected result type {type(result).__name__}")


def main() -> None:
    store = sys.argv[1]
    ds = xr.open_dataset(store, engine="zarr")
    env = {"xr": xr, "np": np, "ds": ds}

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        msg = json.loads(line)
        try:
            result = eval(msg["code"], env)  # noqa: S307 — trusted, our own codegen
            out = {"id": msg["id"], "ok": True, **describe(result)}
        except Exception as exc:  # report, don't die — one bad chain shouldn't kill the run
            out = {"id": msg["id"], "ok": False, "error": f"{type(exc).__name__}: {exc}"}
        sys.stdout.write(json.dumps(out) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: test/differential/oracle.ts
/**
 * Drive the persistent Python {@link file://./oracle.py} oracle over a
 * line-delimited JSON protocol: one request (a transform tail to `eval`) in, one
 * structure-and-signature response out. Kept alive across a whole property run so
 * xarray import + store open are paid once.
 *
 * @module
 */
import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { createInterface, type Interface } from "node:readline";

/** One array's value signature (data variables only; see oracle.py). */
export interface Sig {
  shape: number[];
  sum: number | null;
  first: number | null;
  last: number | null;
}

/** The oracle's description of an eval result. */
export interface OracleResult {
  id: number;
  ok: boolean;
  error?: string;
  kind?: "dataset" | "dataarray";
  name?: string | null;
  dims?: Record<string, number>;
  data_vars?: string[];
  coords?: string[];
  sig?: Sig | Record<string, Sig>;
}

/** A live oracle subprocess. */
export class Oracle {
  readonly #proc: ChildProcessWithoutNullStreams;
  readonly #rl: Interface;
  readonly #pending = new Map<number, (r: OracleResult) => void>();
  #nextId = 0;
  #stderr = "";

  constructor(scriptPath: string, storeDir: string, python = "python") {
    this.#proc = spawn(python, [scriptPath, storeDir]);
    this.#proc.stderr.on("data", (d: Buffer) => {
      this.#stderr += d.toString();
    });
    this.#rl = createInterface({ input: this.#proc.stdout });
    this.#rl.on("line", (line) => {
      if (!line.trim()) return;
      const msg = JSON.parse(line) as OracleResult;
      const cb = this.#pending.get(msg.id);
      if (cb) {
        this.#pending.delete(msg.id);
        cb(msg);
      }
    });
  }

  /** Evaluate one transform tail (e.g. `ds.isel(time=0)`) and resolve with its structure. */
  eval(code: string): Promise<OracleResult> {
    const id = this.#nextId++;
    return new Promise((resolve, reject) => {
      const onExit = () =>
        reject(new Error(`oracle exited before replying to #${id}. stderr:\n${this.#stderr}`));
      this.#proc.once("exit", onExit);
      this.#pending.set(id, (r) => {
        this.#proc.removeListener("exit", onExit);
        resolve(r);
      });
      this.#proc.stdin.write(`${JSON.stringify({ id, code })}\n`);
    });
  }

  /** Captured stderr, for diagnosing a crash. */
  get stderr(): string {
    return this.#stderr;
  }

  /** Shut the subprocess down. */
  close(): void {
    this.#rl.close();
    this.#proc.stdin.end();
    this.#proc.kill();
  }
}
